articles/introduction-to-numpy/index.md

```diff
@@ -0,0 +1,242 @@
+
+print(matrix_A)
+
+"""
+Output - [[1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4]]
+"""
+#Loop over every element in the matrix
+for i in range(4):
+    for j in range(i,4):
+#Swap the elements
+        matrix_A[i][j], matrix_A[j][i] = matrix_A[j][i], matrix_A[i][j]
+    
+print(matrix_A)
+
+"""
+Output - [[1, 1, 1, 1],
+          [2, 2, 2, 2],
+          [3, 3, 3, 3],
+          [4, 4, 4, 4]]
+"""
+```
+
+How does this work out in NumPy? Well, it’s pretty simple.
+
+``` python3
+#Initialize Matrix to a set of values
+matrix_A = [[1,2,3,4] for _ in range(4)]
+
+print(matrix_A)
+
+"""
+Output - [[1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4],
+          [1, 2, 3, 4]]
+"""
+#Convert it into a NumPy array
+np_matrix_A = np.array(matrix_A)
+
+#And transpose!
+np_matrix_A = np_matrix_A.T
+"""
+Output - [[1, 1, 1, 1],
+          [2, 2, 2, 2],
+          [3, 3, 3, 3],
+          [4, 4, 4, 4]]
+"""
+        
+
+```
+        
+You can find more such functions at [https://numpy.org/doc/](https://numpy.org/doc/)
+
+With all that being said, let’s address the elephant in the room.
+
+
+#### How is NumPy so fast?
+Let us analyse the example where we transposed a matrix. One key point to remember is – in any scripting language, a major performance dipper is the use of unnecessary for loops, especially when it is used to perform a single operation (in this case, swapping two elements) to each element on a large dataset.
+Upon crunching a few numbers, transposing a (10000 x 10000) matrix using for loops takes 58.8596s and NumPy takes significantly lesser time. The reason behind such high performance is a tiny concept called vectorization that NumPy implements, wherein it groups element-wise operations together. Such a vectorized approach applies to all elements in an array.
+![Vectorized](/engineering-education/introduction-to-nummpy/Vectorized.jpg)
+ 
+This is the under-the-hood reason why NumPy’s calculations are off the charts.
+Not to mention, the overhead of making sure the memory blocks line up correctly before pouring a ndarray into a C - compiled NumPy function will overwhelm any time benefit if the array isn't relatively large. When compared the performance of NumPy arrays vs Python lists as the number of elements increases is shown in the chart, with the breakeven size observed to be around 200 elements.
+ ![Graph](/engineering-education/introduction-to-nummpy/numpyGraph.jpg)
+Like all things come full circle, let’s try to wrap this up with a performance comparison to our first example – multiplying every element of an array by 5.
+But this time, the NumPy way.
+``` python3
+import time as time
+#Create a NumPy array having 100000 elements
+list_a = np.array([i for i in range(100000)])
+
+start = time.time()
+
+#Multiply by 5 the vectorized way
+list_b = list_a * 5
+
+print(time.time() - start)
+#Output - 0.0004646778106689453        
+
+```
+
+
+#### TL; DR
```

Comment:
Section articles usually follow the convention of having the last paragraph as a conclusion, so it would be better to rename this to conclusion